### PR TITLE
5627: Update organisation PUT endpoint

### PR DESCRIPTION
**Pull request by assoomro**

### CODE REVIEW

File: src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs [hunk 1]
```diff
@@ -84,5 +84,25 @@ public async Task<ActionResult> GetByIdAsync(Guid id)
...
             });
         }
+
+        [HttpPut]
+        [Route("{id}")]
+        public async Task<ActionResult> UpdateOrganisationByIdAsync(Guid id, UpdateOrganisationViewModel viewModel)
+        {
+            if (viewModel is null)
+                throw new ArgumentNullException(nameof(viewModel));
+
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs:94]: Extra newline

File: src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs [hunk 2]
```diff
@@ -84,5 +84,25 @@ public async Task<ActionResult> GetByIdAsync(Guid id)
                 }
             });
         }
+
+        [HttpPut]
+        [Route("{id}")]
+        public async Task<ActionResult> UpdateOrganisationByIdAsync(Guid id, UpdateOrganisationViewModel viewModel)
+        {
+            if (viewModel is null)
+                throw new ArgumentNullException(nameof(viewModel));
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs:93]: Might be worth sticking to what we have done throught and use { } for if statements

File: src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OrganisationRepository.cs [hunk 3]
```diff
@@ -26,5 +26,14 @@ public async Task<Organisation> GetByIdAsync(Guid id)
         {
             return await _context.Organisations.FirstOrDefaultAsync(org => org.OrganisationId == id);
         }
+
+        public async Task UpdateAsync(Organisation organisation)
+        {
+            if (organisation is null)
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OrganisationRepository.cs:32]: Not sure if this check is needed as you are checking if it is null before-hand.

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Update.5176.feature [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+@5176
+Scenario: 4. Service Failure
+	Given a user is logged in
+		| Username          | Password       | Scope        |
+		| authority@doe.com | Str0nkP4s5w0rd | Organisation |
+	And the call to the database will fail
+	When  a PUT request is made to update an organisation with name Organisation 1
+		| CatalogueAgreementSigned |
+		| false                    |
+	Then a response with status code 500 is returned
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Update.5176.feature:56]: New line at the end of the feature file, to keep it consistant

File: tests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests/Controllers/OrganisationControllerTests.cs [hunk 5]
```diff
@@ -209,5 +209,65 @@ public async Task GetIdByAsync_VerifyMethodIsCalledOnce_VerifiesMethod()
...
+        {
+            using var controller = OrganisationControllerBuilder.Create().WithUpdateOrganisation(new Organisation()).Build();
+
+            var response = await controller.UpdateOrganisationByIdAsync(Guid.Empty, new UpdateOrganisationViewModel());
+
+            response.Should().BeOfType<NoContentResult>();
+        }
+
+        [Test]
+        public async Task UpdateOrganisationByIdAsync_UpdateOrganisation_ReturnsStatusNotFound()
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests/Controllers/OrganisationControllerTests.cs:224]: Maybe add a bit more in the method name, as at the moment it looks like you are doing the same thing (as the method above) but returning different things

File: src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs (hunk 1 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs:94]: 👍

File: src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs (hunk 2 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs:93]: I think this was raised in the code standards meeting but we didn't reach a conclusion. 
Although I prefer 1 line statements this way, I don't mind compromising here. I'll add the braces :]

File: src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OrganisationRepository.cs (hunk 3 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OrganisationRepository.cs:32]: I was doing the check with future uses in mind (this method might get used somewhere else and I can't guarantee that the user will check for null). If you still feel like that's being overly cautious let me know and I'll change it :]

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Update.5176.feature (hunk 4 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Update.5176.feature:56]: 👍

File: tests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests/Controllers/OrganisationControllerTests.cs (hunk 5 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Organisations.Api.UnitTests/Controllers/OrganisationControllerTests.cs:224]: 👍

File: src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs (hunk 2 above)
Comment (bjss-chris) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs:93]: The code standards decision was developer's prerogative on this one.

File: src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs [hunk 6]
```diff
@@ -84,5 +84,25 @@ public async Task<ActionResult> GetByIdAsync(Guid id)
                 }
             });
         }
+
+        [HttpPut]
+        [Route("{id}")]
+        public async Task<ActionResult> UpdateOrganisationByIdAsync(Guid id, UpdateOrganisationViewModel viewModel)
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs:90]: `viewModel` isn't actually a view model here but there are currently inconsistencies and differences of opinion on this one so this is something we're planning to discuss tomorrow.

File: src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs (hunk 2 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs:93]: I remember wrongly then 😄
I'm not that fussed about this, plus, I've already added the brackets in.

File: src/NHSD.BuyingCatalogue.Organisations.Api/ViewModels/Organisations/UpdateOrganisationViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+﻿namespace NHSD.BuyingCatalogue.Organisations.Api.ViewModels.Organisations
+{
+    public sealed class UpdateOrganisationViewModel
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Organisations.Api/ViewModels/Organisations/UpdateOrganisationViewModel.cs:3]: As above, this isn't actually a view model but there are currently inconsistencies and differences of opinion on this one so this is something we're planning to discuss tomorrow.

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Update.5176.feature [hunk 8]
```diff
@@ -0,0 +1,56 @@
+﻿Feature: Edit Buyer Organisation
+    As a Authority User
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Update.5176.feature:2]: As *an* authority user

File: src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs (hunk 6 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Controllers/OrganisationsController.cs:90]: I see. For me, in this instance a VM is a DTO representing the json we'd be getting / returning.
I don't mind renaming this, as long as its solution-wide, agreed upon and understood by the whole team. But more on this tomorrow.

I'll be leaving it as is for now, and based on the outcomes of tomorrow's meeting either change it or keep it.

File: src/NHSD.BuyingCatalogue.Organisations.Api/ViewModels/Organisations/UpdateOrganisationViewModel.cs (hunk 7 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/ViewModels/Organisations/UpdateOrganisationViewModel.cs:3]: Reply as above.

File: src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OrganisationRepository.cs (hunk 3 above)
Comment (RuthBerry96) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Repositories/OrganisationRepository.cs:32]: No you are perfectly right tbh.

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Update.5176.feature (hunk 8 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Update.5176.feature:2]: 👍

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Update.5176.feature [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+	Given Organisations exist
+		| Name           | OdsCode | PrimaryRoleId | CatalogueAgreementSigned | Line1 | Line2      | Line3        | Line4       | Town  | County         | Postcode | Country |
+		| Organisation 1 | Ods 1   | ID 1          | true                     | 12    | Brick Lane | Central Area | City Centre | Leeds | West Yorkshire | LS1 1AW  | England |
+	And Users exist
+		| Id  | OrganisationName | FirstName | LastName | Email             | PhoneNumber | Disabled | Password       | OrganisationFunction |
+		| 123 | Organisation 1   | John      | Doe      | authority@doe.com | 01234567890 | false    | Str0nkP4s5w0rd | Authority            |
+		| 234 | Organisation 1   | Jane      | Doe      | buyer@doe.com     | 01234567890 | false    | W3AkP4s5w0rd   | Buyer                |
+
+@5176
+Scenario: 1. A authority user can update organisations Catalogue Agreement Status
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Update.5176.feature:16]: *An* authority user...

Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Update.5176.feature:16]: 👍

**Pull request merged**